### ISSUE TITLE
_update API might fail if values are complex objects

### ISSUE THREAD

**Issue by iverase** [elastic/elasticsearch#90726]
The way the _update API merges values from complex objects might lead to inconsistent results. This can easily be seen in geo_shape fields when using geoJson. For example if we create a GeoJson field:

```
PUT test
{
  "mappings": {
    "properties": {
      "location": {
        "type": "geo_shape"
      }
    }
  }
}
```

And we index a point:

```
POST test/_doc/0
{
  "location" : {"type":"Point","coordinates" : [0,0]}
}
```

If we try to update it with a geometry collection:

```
POST test/_update/0
{
  "doc":{"location": {"type":"GeometryCollection","geometries":[{"type":"Point","coordinates":[0,0]}]}}
}
```
We get an error like:

```
{
  "error": {
    "root_cause": [
      {
        "type": "mapper_parsing_exception",
        "reason": "failed to parse field [location] of type [geo_shape]"
      }
    ],
    "type": "mapper_parsing_exception",
    "reason": "failed to parse field [location] of type [geo_shape]",
    "caused_by": {
      "type": "x_content_parse_exception",
      "reason": "Failed to build [geojson] after last required field arrived",
      "caused_by": {
        "type": "parse_exception",
        "reason": "parameter coordinates is not supported for type GeometryCollection"
      }
    }
  },
  "status": 400
}
```

This is because when merging the old and the new value, the coordinates field from the original point is not deleted.

**Comment by javanna** [comment]
I believe this is due to how the Update API works. It takes the existing _source and it merges it with the provided _source, recursively. This works well in most of the cases, but I can see how it causes problems for field types that support parsing objects natively like the example above. I think that the root of the problem is that the merge operation is not mappings-aware, yet I am not too sure it should be. Is using a scripted update an option here instead of providing a partial document? That should help have more control over what happens with each inner field. I will re-label as this does not look like a search problem to me.

**Comment by iverase** [comment]
I agree that this issue is because of how the API works and making it mapping aware might no be a good idea. I think we should at least document it and provide some guidelines, e.g as you suggested using a scripted update.

(Another ugly option is to perform two updates: the first update will nullify the value on the field and the second one will update to the new value).
